cli/azd/pkg/containerregistry/remote_build.go

```diff
@@ -0,0 +1,217 @@
+	}
+
+	regClient, err := armcontainerregistry.NewRegistriesClient(subscriptionID, cred, r.armClientOptions)
+	if err != nil {
+		return armcontainerregistry.SourceUploadDefinition{}, err
+	}
+
+	sourceUploadRes, err := regClient.GetBuildSourceUploadURL(ctx, resourceGroupName, registryName, nil)
+	if err != nil {
+		return armcontainerregistry.SourceUploadDefinition{}, err
+	}
+
+	blobClient, err := blockblob.NewClientWithNoCredential(*sourceUploadRes.UploadURL, nil)
+	if err != nil {
+		return armcontainerregistry.SourceUploadDefinition{}, err
+	}
+
+	dockerContext, err := os.Open(buildSourcePath)
+	if err != nil {
+		return armcontainerregistry.SourceUploadDefinition{}, err
+	}
+	defer dockerContext.Close()
+
+	_, err = blobClient.UploadFile(context.Background(), dockerContext, nil)
+	if err != nil {
+		return armcontainerregistry.SourceUploadDefinition{}, err
+	}
+
+	return sourceUploadRes.SourceUploadDefinition, nil
+}
+
+// RunDockerBuildRequestWithLogs initiates a remote build on the specified registry and streams the logs to the provided
+// writer.
+func (r *RemoteBuildManager) RunDockerBuildRequestWithLogs(
+	ctx context.Context,
+	subscriptionID, resourceGroupName, registryName string,
+	buildRequest *armcontainerregistry.DockerBuildRequest,
+	writer io.Writer,
+) error {
+	tracing.IncrementUsageAttribute(fields.RemoteBuildCount.Int(1))
+
+	cred, err := r.credentialProvider.CredentialForSubscription(ctx, subscriptionID)
+	if err != nil {
+		return err
+	}
+
+	regClient, err := armcontainerregistry.NewRegistriesClient(subscriptionID, cred, r.armClientOptions)
+	if err != nil {
+		return err
+	}
+
+	runPoller, err := regClient.BeginScheduleRun(ctx, resourceGroupName, registryName, buildRequest, nil)
+	if err != nil {
+		return err
+	}
+
+	resp, err := runPoller.Poll(ctx)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	var runResp armcontainerregistry.RegistriesClientScheduleRunResponse
+	if err := json.NewDecoder(resp.Body).Decode(&runResp); err != nil {
+		return err
+	}
+
+	runClient, err := armcontainerregistry.NewRunsClient(subscriptionID, cred, r.armClientOptions)
+	if err != nil {
+		return err
+	}
+
+	logRes, err := runClient.GetLogSasURL(ctx, resourceGroupName, registryName, *runResp.Properties.RunID, nil)
+	if err != nil {
+		return err
+	}
+
+	logBlobClient, err := blockblob.NewClientWithNoCredential(*logRes.LogLink, &blockblob.ClientOptions{
+		ClientOptions: r.armClientOptions.ClientOptions,
+	})
+	if err != nil {
+		return err
+	}
+
+	var buildLog bytes.Buffer
+
+	err = streamLogs(ctx, logBlobClient, io.MultiWriter(&buildLog, writer))
+	if err != nil {
+		return err
+	}
+
+	runRes, err := runClient.Get(ctx, resourceGroupName, registryName, *runResp.Properties.RunID, nil)
+	if err != nil {
+		return err
+	}
+
+	if *runRes.Properties.Status != armcontainerregistry.RunStatusSucceeded {
+		return fmt.Errorf("remote build failed: %v", buildLog.String())
+	}
```

Comment:
I wonder if it is guaranteed that when the blob is marked as completed, the run status has also been finalized.

I hit a weird one-time build failure here:
![image](https://github.com/user-attachments/assets/89b54ab6-b7f3-47f3-b300-ad3110f19387)
![image](https://github.com/user-attachments/assets/05b9aa94-fc68-40cd-8600-998b3db92497)

The last line of the log "Run ID: ch3 was successful after 41s" seems to indicate it's successful, but `azd` had failed it as `remote build failed`.

Unfortunately, without knowing the intended spec of the usage of the API, I'm not sure what the right thing to do is. The easiest thing we could do is to update the logic to poll for completion here. This works optimally if the blob is guaranteed to be marked as Complete prior to the operation completion.

A different alternative perhaps is to always block and poll on the operation, while streaming logs in a separate goroutine (similar to the ARM deployments). And perhaps in this model, we'd add a timeout for how long the operation would wait for logs to be finalized before giving up and emitting the total output.

Reply:
> A different alternative perhaps is to always block and poll on the operation, while streaming logs in a separate goroutine (similar to the ARM deployments). And perhaps in this model, we'd add a timeout for how long the operation would wait for logs to be finalized before giving up and emitting the total output.

Yeah - I think I should just do that.  I'll make the change.

Reply:
> Unfortunately, without knowing the intended spec of the usage of the API, I'm not sure what the right thing to do is. The easiest thing we could do is to update the logic to poll for completion here. This works optimally if the blob is guaranteed to be marked as Complete prior to the operation completion.

Ended up doing this and made some comments as to why we can't use `PollUntilDone` from the poller returned by the initial call to enqueue the run.